src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -0,0 +1,1146 @@
+                    String roleName = entry.getKey();
+                    RoleV7 role = entry.getValue();
+
+                    roleSetBuilder.next(roleName);
+
+                    for (RoleV7.Index indexPermissions : role.getIndex_permissions()) {
+                        ImmutableSet<String> permissions = actionGroups.resolve(indexPermissions.getAllowed_actions());
+
+                        if (indexPermissions.getIndex_patterns().contains("*")) {
+                            // Wildcard index patterns are handled in the static IndexPermissions object.
+                            continue;
+                        }
+
+                        WildcardMatcher indexMatcher = IndexPattern.from(indexPermissions.getIndex_patterns()).getStaticPattern();
+
+                        if (indexMatcher == WildcardMatcher.NONE) {
+                            // The pattern is likely blank because there are only templated patterns.
+                            // Index patterns with templates are not handled here, but in the static IndexPermissions object
+                            continue;
+                        }
+
+                        for (String permission : permissions) {
+                            WildcardMatcher actionMatcher = WildcardMatcher.from(permission);
+                            Collection<String> matchedActions = actionMatcher.getMatchAny(wellKnownIndexActions, Collectors.toList());
+
+                            for (Map.Entry<String, IndexAbstraction> indicesEntry : indexMatcher.iterateMatching(
+                                indices.entrySet(),
+                                Map.Entry::getKey
+                            )) {
+                                for (String action : matchedActions) {
+                                    CompactMapGroupBuilder.MapBuilder<
+                                        String,
+                                        DeduplicatingCompactSubSetBuilder.SubSetBuilder<String>> indexToRoles = actionToIndexToRoles
+                                            .computeIfAbsent(action, k -> indexMapBuilder.createMapBuilder());
+
+                                    indexToRoles.get(indicesEntry.getKey()).add(roleName);
+
+                                    if (indicesEntry.getValue() instanceof IndexAbstraction.Alias) {
+                                        // For aliases we additionally add the sub-indices to the privilege map
+                                        for (IndexMetadata subIndex : indicesEntry.getValue().getIndices()) {
+                                            indexToRoles.get(subIndex.getIndex().getName()).add(roleName);
+                                        }
+                                    }
+
+                                    if (roleSetBuilder.getEstimatedByteSize() + indexMapBuilder
+                                        .getEstimatedByteSize() > statefulIndexMaxHeapSize.getBytes()) {
+                                        log.info(
```

Comment:
codecov is showing that these lines are not hit. Are there any tests that cover this scenario?

Reply:
Added in https://github.com/opensearch-project/security/pull/4380/commits/05fac0792d7b0607f1f0f2aa4cb1f29d821cf01b